File: app/src/main/java/org/tvheadend/tvhclient/ui/features/epg/EpgViewPagerRecyclerViewAdapter.java
package org.tvheadend.tvhclient.ui.features.epg;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;

import org.tvheadend.tvhclient.R;
import org.tvheadend.tvhclient.domain.entity.EpgChannel;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.RecyclerView;

class EpgViewPagerRecyclerViewAdapter extends RecyclerView.Adapter<EpgViewPagerViewHolder> implements Filterable {

    private final float pixelsPerMinute;
    private final long startTime;
    private final long endTime;
    private final FragmentActivity activity;
    private final RecyclerView.RecycledViewPool viewPool;
    private final List<EpgChannel> channelList = new ArrayList<>();
    private List<EpgChannel> channelListFiltered = new ArrayList<>();

    EpgViewPagerRecyclerViewAdapter(FragmentActivity activity, float pixelsPerMinute, long startTime, long endTime) {
        this.activity = activity;
        this.pixelsPerMinute = pixelsPerMinute;
        this.startTime = startTime;
        this.endTime = endTime;
        this.viewPool = new RecyclerView.RecycledViewPool();
    }

    @NonNull
    @Override
    public EpgViewPagerViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final View view = LayoutInflater.from(parent.getContext()).inflate(viewType, parent, false);
        return new EpgViewPagerViewHolder(activity, view, pixelsPerMinute, startTime, endTime, viewPool);
    }

    @Override
    public void onBindViewHolder(@NonNull EpgViewPagerViewHolder holder, int position) {
        EpgChannel epgChannel = channelListFiltered.get(position);
        holder.bindData(epgChannel);
    }

    @Override
    public int getItemCount() {
        return channelListFiltered != null ? channelListFiltered.size() : 0;
    }

    @Override
    public int getItemViewType(final int position) {
        return R.layout.epg_program_list_adapter;
    }

    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence charSequence) {
                String charString = charSequence.toString();
                if (charString.isEmpty()) {
                    channelListFiltered = channelList;
                } else {
                    List<EpgChannel> filteredList = new ArrayList<>();
                    // Iterate over the available channels. Use a copy on write
                    // array in case the channel list changes during filtering.
                    for (@NonNull EpgChannel channel : new CopyOnWriteArrayList<>(channelList)) {
                        if (channel.getName() != null
                                && channel.getName().toLowerCase().contains(charString.toLowerCase())) {
                            filteredList.add(channel);
                        }
                    }
                    channelListFiltered = filteredList;
                }

                FilterResults filterResults = new FilterResults();
                filterResults.values = channelListFiltered;
                return filterResults;
            }

            @SuppressWarnings("unchecked")
            @Override
            protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
                channelListFiltered = (ArrayList<EpgChannel>) filterResults.values;
                notifyDataSetChanged();
            }
        };
    }

    public void addItems(@NonNull List<EpgChannel> channels) {
        channelList.clear();
        channelListFiltered.clear();
        channelList.addAll(channels);
        channelListFiltered.addAll(channels);
        notifyDataSetChanged();
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/ui/features/epg/EpgViewPagerRecyclerViewAdapter.kt
package org.tvheadend.tvhclient.ui.features.epg

import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView
import org.tvheadend.tvhclient.R
import org.tvheadend.tvhclient.domain.entity.EpgChannel
import java.util.*
import java.util.concurrent.CopyOnWriteArrayList

internal class EpgViewPagerRecyclerViewAdapter(private val activity: FragmentActivity, private val epgViewModel: EpgViewModel, private val fragmentId: Int) : RecyclerView.Adapter<EpgViewPagerViewHolder>(), Filterable {

    private val viewPool: RecyclerView.RecycledViewPool = RecyclerView.RecycledViewPool()
    private val channelList = ArrayList<EpgChannel>()
    private var channelListFiltered: MutableList<EpgChannel> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): EpgViewPagerViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(viewType, parent, false)
        return EpgViewPagerViewHolder(view, activity, epgViewModel, fragmentId, viewPool)
    }

    override fun onBindViewHolder(holder: EpgViewPagerViewHolder, position: Int) {
        val epgChannel = channelListFiltered[position]
        holder.bindData(epgChannel)
    }

    override fun getItemCount(): Int {
        return channelListFiltered.size
    }

    override fun getItemViewType(position: Int): Int {
        return R.layout.epg_program_list_adapter
    }

    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(charSequence: CharSequence): FilterResults {
                val charString = charSequence.toString()
                channelListFiltered = if (charString.isEmpty()) {
                    channelList
                } else {
                    val filteredList = ArrayList<EpgChannel>()
                    // Iterate over the available channels. Use a copy on write
                    // array in case the channel list changes during filtering.
                    for (channel in CopyOnWriteArrayList(channelList)) {
                        if (channel.name != null && channel.name!!.toLowerCase().contains(charString.toLowerCase())) {
                            filteredList.add(channel)
                        }
                    }
                    filteredList
                }

                val filterResults = FilterResults()
                filterResults.values = channelListFiltered
                return filterResults
            }

            override fun publishResults(charSequence: CharSequence, filterResults: FilterResults) {
                @Suppress("UNCHECKED_CAST")
                channelListFiltered = filterResults.values as ArrayList<EpgChannel>
                notifyDataSetChanged()
            }
        }
    }

    fun addItems(channels: List<EpgChannel>) {
        channelList.clear()
        channelListFiltered.clear()
        channelList.addAll(channels)
        channelListFiltered.addAll(channels)
        notifyDataSetChanged()
    }
}
